Darwin gtk2/pygtk quartz backend

johbo opened the pull request:
###### Motivation for this change

I wanted to use a pygtk based application on darwin and noticed that it relied on the X11 backend of gtk which leads to trouble in multi screen setups (lost windows out of screen). Using the quartz backend also leads to a nicer display of the applications.

This Pull Request adjust GTK2 based on the work from #20658. I aimed to keep the changes to the GTK2 Derivation as close as possible to the other PR. The main change is that I added the CoreText dependencies to cairo and pango, so that fonts are actually visible.

Due to #12346 starting the applications still needs a twist by using something like `DYLD_FRAMEWORK_PATH=/System/Library/Frameworks`.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer committed:
gtk-mac-integration: init at 2.0.8

johbo committed:
cairo: Add CoreText for darwin

This helps so that fonts are properly rendered in gtk when used with the backend
"quartz".

johbo committed:
pango: Add darwin frameworks for fond rendering

johbo committed:
librsvg: Add framework ApplicationServices for darwin

johbo committed:
gtk2: Use quartz backend on darwin

johbo committed:
pygtk: Add flag -ObjC for darwin

Had to add this so that it was able to parse the headers from the frameworks.

johbo commented:
Regarding the binaries I've tested `gtk-demo` and `pygtk-demo`, those worked fine.

johbo committed:
cairo: Add ApplicationServices as propagatedBuildInput

Noticed that it caused depending packages to fail, like librsvg, imagemagiv, graphviz.

johbo commented:
I've done some more experiments, it appears to me as if ApplicationServices should go into the propagatedBuildInputs of cairo.

vbgl commented:
This breaks `gnome2.gtksourceview`. It now reports the following error.

```
checking for IGE_MAC... no
configure: error: Package requirements (ige-mac-integration) were not met:

No package 'ige-mac-integration' found
```

You may need to package some extra libraries (https://github.com/rhult/ige-mac-integration).

johbo commented:
Uh, it might cause a few more issues like that I guess. I'll try to add ige-mac-integration and make things work.

johbo committed:
gtksourceview: Add gtk-mac-integration for darwin

Had to pick two commits as patches from the repository which update the
dependency ige-mac-integration to gtk-mac-integration.

johbo commented:
@vbgl think I got the gtksourceview issue figured out. Do you have any suggestions how I could best check how much this change might affect other packages?

vbgl commented:
Indeed, `gtksourceview` now builds. But this PR breaks `coqide`, `inkscape`, `zim`: they build but crash on startup.

LnL7 commented:
@johbo nox-review and https://github.com/NixOS/nixpkgs/issues/19045#issuecomment-256920869

johbo commented:
Hmm, it seems that all gtk2 based applications will suffer from https://github.com/NixOS/nixpkgs/issues/12346 until we get that sorted. I wonder if it would be better to make this gtk2 change opt-in and build by default for X11 until that's solved.

matthewbauer commented on pkgs/desktops/gnome-2/desktop/gtksourceview/default.nix:
```diff
@@ -9,6 +12,21 @@ stdenv.mkDerivation rec {
-    gettext];
+
+  patches = optionals stdenv.isDarwin [
+    ./0001-Change-IgeMacIntegration-to-GtkOSXApplication.patch
```
Can you use fetchpatch for these?

johbo replied:
Thanks for the hint, was not yet aware of it. I'll try to use it for both patches.

matthewbauer commented on pkgs/development/libraries/librsvg/default.nix:
```diff
@@ -1,5 +1,5 @@
 { lib, stdenv, fetchurl, pkgconfig, glib, gdk_pixbuf, pango, cairo, libxml2, libgsf
-, bzip2, libcroco, libintlOrEmpty
+, bzip2, libcroco, libintlOrEmpty, darwin
```
This shouldn't be needed.

matthewbauer reviewed (changes requested):
These changes look good! It probably should be merged into staging first though.

Once this gets into master I can work on merging #20658 again.

johbo commented:
I'll rebase this quickly

matthewbauer commented:
cc @lnl7 @copumpkin

LnL7 commented:
Looks like about ~1400 packages will rebuild, I think that's still ok to do on master.
